### ISSUE TITLE
Create an IDatabaseFacade interface for mocking purposes in business logic

### ISSUE THREAD

**Issue by LaurentLardenois** [dotnet/efcore#28397]
I have a piece of business logic that goes as follows (Obfuscated pseudo code because I cannot share it literally)

public async Task SomeFunction(CancelationToken ct)
{
    await DoSomeCallsOnExternalService1(ct);
    var transaction = await _injectedContext.Database.BeginTransactionAsync(ct);
    await DoSomeDatabaseStuff();
    try
    {
        await DoSomeCallsOnExternalService2(ct);
    }
    catch (ApiException e)
    {
        Logger.Error(e);
        await transaction.RollbackAsync(ct);
        await RollbackCallsOnExternalService1(ct);
    }
}

This is a crucial piece of business logic where unit testing, if the rollback is correctly invoked, is very important. Sadly enough, there are EF calls in it. Why does that make me sad? Because having anything EF related in business logic that you will need to unit test is a one way ticket to hell.
For my DbSets, I have a library that makes unit testing my interactions with them bearable, but even then it is not possible to verify the mock calls that get made due to the huge callstack that the library creates in memory. (It is not really "mocking" the interactions, but it takes away the burden of the huge setup that is required).
For the DatabaseFacade (which I need to call the BeginTransaction) there is a similar issue, namely that there is no interface wrapped around it which makes unit testing the above functionality very difficult. I could make a full in memory database based around a seperate DbContext to get the end result I need, problem is, that goes far beyond the scope of a unit test (and yes, it is a unit test - I am not trying to interact with the database, I am merely trying to see if the calls I want to make are invoked when I want them to be invoked). This might seem obvious, but as I looked around on the internet for possible solutions most of the responses boiled down to either 'You are not unit testing but integration testing' or 'Just make a repository interface to proxy the calls'. This second response doesn't solve the problem, because it forces the developer to proxy all the calls they want to make which is a lot of work. Work that could easily be handed off to (you guessed it) an interface. (There are already issues open about the DbContext not having an interface so I won't repeat that discussion here).

I would like to see the EF project use more interfaces in general, but since there is no request for a specific IDatabaseFacade interface here it is. I understand that it will not be included in the existing EF releases, but I hope my explanation has shown that it would be of great added value to the framework, and gets it on the roadmap for a future EF release (EF 7/8). And, if I have not yet convinced you for this feature, I am looking forward to the discussion that will ensue in this thread.

**Comment by roji** [comment]
What's stopping you from mocking DatabaseFacade as it is? It's an unsealed, public class - mocking isn't just for interfaces.

**Comment by roji** [comment]
@LaurentLardenois I'm assuming you're already mocking your DbContext as well, why not pass that to the DatabaseFacade constructor?

**Comment by LaurentLardenois** [comment]
Without making a stub class for the DbContext, you cant create a proper mock for that either. The issue of this ticket is not that it is impossible to get the tests to work. The issue is that EF has no decent support for testing and puts the developers that want to mock EF functions in a pickle where they either can't test the functionality the want to test, or have to create a wrapping repository pattern around the EF functions to be able to mock them, or have to create full stack stubs, in which case you are no longer unit testing but integration testing. Testing with EF should not take longer or be more complicated than testing any with any other Api, currently it is, that's irrefutable, if it wasn't there would't a ton of hacky workarounds floating around the internet. This issue wasn't opened as a "Please fix my current situation", because I know you won't and can't. This issue is meant to shine light on the problems with EF mocking. If you do have a proper example of how to mock everything I will gladly give it a try, but if it consists out of having to make extra classes solely for test purposes, you are only showing me that EF is indeed not yet test and mock friendly.

**Comment by roji** [comment]
You indeed would need to mock DbContext, but that should be pretty trivial and necessary anyway for other purposes. I wouldn't consider that requirement as making EF mock-unfriendly.

Stepping back, we're indeed aware that in some cases doing mocking with EF Core sometimes isn't as easy as it could be (see #27110 for example). One reason why we haven't invested much in this area, is that mocking doesn't work well with LINQ querying, which relies on composing static function operators on top of the query root (DbSet in the EF case). While it's possible to replace the DbSet with an in-memory collection - or configure EF Core with the InMemory provider - that provides no confidence that the compose LINQ operators behave in the same way (or work at all) as they would on the real database.

As a result, at least for querying, our recommendation is indeed to either test against your real database (integration testing), or to wrap EF Core in a repository which can be mocked. This isn't because we don't want to ber mock-friendly, but because LINQ querying really does present some unique problems. I recommend reading our [testing docs](https://docs.microsoft.com/en-us/ef/core/testing/choosing-a-testing-strategy), which go into this in some detail.

Having said that, it's true that for activities other than querying, EF Core isn't really different from other APIs, and I agree it should support mocking in a seamless way.

**Comment by LaurentLardenois** [comment]
That explains a lot, thank you very much for that clarification!

**Comment by ilmax** [comment]
@LaurentLardenois another option is to not wrap an external service call in a database transaction and just eventually rollback the db changes if the external service call fails. I.e. model your workflow as a saga.
This way you don't have to mock the database facade and you keep your transaction short lived which is generally a good thing.
I know you are probably aware of this but I decided to mention it just in case :)

**Comment by roji** [comment]
@ilmax yeah, that's one of the ways of writing tests which run against a real database (aka integration tests), the various options for that approach are detailed on [this doc page](https://docs.microsoft.com/en-us/ef/core/testing/testing-with-the-database).

**Comment by LaurentLardenois** [comment]
Yeah that is not an option though due to business rule order constraints that are very specific and important for this flow. Would have been a good suggestion otherwise though!
